### PR TITLE
Remove shard-level injector

### PR DESCRIPTION
**Pull request by s1monw**
Today we use a hierarchical injector on the shard level for each shard
created. This commit removes the shard level injector and replaces
it with good old constructor calls. This also removes all shard level plugin
facilities such that plugins can only have node or index level modules.
For plugins that need to track shard lifecycles they should use the relevant
callback from the lifecycle we already provide.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/PluginsService.java [hunk 1]
```diff
@@ -250,14 +250,6 @@ public Settings updatedSettings() {
...
 
-    public Collection<Module> shardModules(Settings indexSettings) {
-        List<Module> modules = new ArrayList<>();
-        for (Tuple<PluginInfo, Plugin> plugin : plugins) {
-            modules.addAll(plugin.v2().shardModules(indexSettings));
-        }
-        return modules;
-    }
-
     public Collection<Class<? extends Closeable>> shardServices() {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/PluginsService.java]: these can go too?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardModule.java [hunk 2]
```diff
@@ -39,8 +34,7 @@
     private final Settings settings;
     private final boolean primary;
 
-    // pkg private so tests can mock
-    Class<? extends EngineFactory> engineFactoryImpl = InternalEngineFactory.class;
+
 
     public IndexShardModule(ShardId shardId, boolean primary, Settings settings) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardModule.java]: this can go?

**Comment by rjernst** [comment]
Fantastic!! LGTM!

**Comment by mikemccand** [comment]
LGTM!

**Comment by s1monw** [comment]
@rjernst @mikemccand pushed some updates

**Comment by rjernst** [comment]
LGTM

**Comment by rjernst** [comment]
New commit still LGTM

**Comment by mikemccand** [comment]
LGTM

**Pull request merged**